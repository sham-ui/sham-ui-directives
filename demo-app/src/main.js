import { createDI, start } from 'sham-ui';
import mainInitializer from './initializers/main';

if ( module.hot ) {
    if ( window.DI ) {
        const store = window.DI.resolve( 'sham-ui:store' );
        const app = store.byId.get( 'app' );
        if ( undefined !== app ) {
            app.remove();
            store.byId.values().forEach( component => {
                try {
                    component.remove();
                } catch ( e ) {
                    // eslint-disable-next-line no-empty
                }
            } );
        }
    }
    module.hot.accept();
}

const DI = createDI();

window.DI = DI;

mainInitializer( DI );
start( DI );
