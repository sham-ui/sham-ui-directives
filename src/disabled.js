/**
 * Directive for "disabled" attribute
 * @example
 * <template>
 *   <input :disabled={{inputDisabled}}/>
 * </template>
 * <script>
 *     import { disabled } from 'sham-ui-directives';
 *     import { options } from 'sham-ui-macro/babel.macro';
 *
 *     class Input extends Template {
 *         \@options inputDisabled = true;
 *
 *         constructor() {
 *             super( ...arguments );
 *             this.directives = {
 *                 ...( this.directives || {} ),
 *                 disabled
 *             };
 *         }
 *     }
 *
 *     export default Input;
 * </script>
 */
export default class disabled {
    constructor() {
        this.node = null;
    }

    bind( node ) {
        this.node = node;
    }

    unbind() {
        this.node = null;
    }

    update( value ) {
        this.node.disabled = value;
    }
}
