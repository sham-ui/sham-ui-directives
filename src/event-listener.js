class EventListener {
    constructor( type ) {
        this.type = type;
        this.handler = null;
        this.callback = this.callback.bind( this );
    }

    callback( event ) {
        this.handler( event );
    }

    bind( node ) {
        node.addEventListener( this.type, this.callback );
    }

    unbind( node ) {
        node.removeEventListener( this.type, this.callback );
    }

    update( handler ) {
        this.handler = handler;
    }
}

export default {
    onclick: class extends EventListener {
        constructor() {
            super( 'click' );
        }
    },
    onsubmit: class extends EventListener {
        constructor() {
            super( 'submit' );
        }
    },
    onload: class extends EventListener {
        constructor() {
            super( 'load' );
        }
    },
    onresize: class extends EventListener {
        constructor() {
            super( 'resize' );
        }
    },
    onscroll: class extends EventListener {
        constructor() {
            super( 'scroll' );
        }
    },
    oninput: class extends EventListener {
        constructor() {
            super( 'input' );
        }
    },
    onselect: class extends EventListener {
        constructor() {
            super( 'select' );
        }
    }
};
