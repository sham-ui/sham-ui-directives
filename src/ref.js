/**
 * Directive for reference to Node
 * @example
 * <template>
 *   <input :ref="inputField" value="default value"/>
 * </template>
 * <script>
 *     import { ref } from 'sham-ui-directives';
 *
 *     class Input extends Template {
 *         constructor() {
 *             super( ...arguments );
 *             this.directives = {
 *                 ...( this.directives || {} ),
 *                 ref
 *             };
 *         }
 *
 *         didMount() {
 *             alert( this.inputField.value );
 *         }
 *     }
 *
 *     export default Input;
 * </script>
 */
export default class ref {
    constructor( component ) {
        this.component = component;
        this.node = null;
        this.name = null;
    }

    bind( node ) {
        this.node = node;
    }

    unbind() {
        this.node = null;
        if ( this.name ) {
            delete this.component[ this.name ];
        }
        this.name = null;
    }

    update( name ) {
        this.name = name;
        this.component[ name ] = this.node;
    }
}
