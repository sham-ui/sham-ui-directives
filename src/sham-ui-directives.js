import ShamUiDirectives from './sham-ui-directives.sht';

export default class extends ShamUiDirectives {

}
