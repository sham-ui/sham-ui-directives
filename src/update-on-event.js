
/**
 * Base directive class for event listener
 * @example
 * import { UpdateOnEvent } from 'sham-ui-directives';
 *
 * // directive for update on 'input' event listener
 * class updateOnEvent extends UpdateOnEvent {
 *     constructor() {
 *         super( 'input' );
 *     }
 * }
 */
export class UpdateOnEvent {
    constructor( component, type ) {
        this.type = type;
        this.callback = this.callback.bind( this );
        this.component = component;
    }

    callback() {
        this.component.update( {
            [ this.field ]: this._getValue()
        } );
    }

    bind( node ) {
        this.node = node;
        node.addEventListener( this.type, this.callback );
    }

    unbind( node ) {
        node.removeEventListener( this.type, this.callback );
        this.node = null;
    }

    update( args ) {
        this.field = args;
    }

    _getValue() {
        const node = this.node;
        if ( node.type === 'checkbox' ) {
            return node.checked;
        }
        return node.value;
    }
}

/**
 * Update on 'change' event
 */
export  class onChangeUpdate extends UpdateOnEvent {
    constructor( component ) {
        super( component, 'change' );
    }
}

/**
 * Update on 'input' event
 */
export  class onInputUpdate extends UpdateOnEvent {
    constructor( component ) {
        super( component, 'input' );
    }
}
