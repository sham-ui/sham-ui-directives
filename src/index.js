import disabled from './disabled';
import events from './event-listener';

export default {
    disabled,
    ...events
};
