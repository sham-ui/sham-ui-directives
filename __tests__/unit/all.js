import renderer from 'sham-ui-test-helpers';
import Dummy from './events/Dummy.sht';
import * as directives from '../../src/index';

it( 'click', () => {
    expect.assertions( 2 );

    const click = jest.fn();
    const meta = renderer( Dummy, {
        click
    }, {
        directives
    } );
    expect( meta.toJSON() ).toMatchSnapshot();
    document.querySelector( 'button' ).click();
    expect( click ).toHaveBeenCalledTimes( 1 );
} );
