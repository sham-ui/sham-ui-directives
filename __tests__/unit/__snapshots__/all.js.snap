// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`click 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "click": [MockFunction],
    "directives": Object {
      "EventListener": [Function],
      "disabled": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
      "ref": [Function],
    },
  },
  "html": "<button>click me</button>",
}
`;
